rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -233,7 +233,7 @@ NodeParameters::set_parameters_atomically(
 
   std::swap(tmp_map, parameters_);
   parameter_event->stamp = node_clock_->get_clock()->now();
-  events_publisher_->publish(parameter_event);
+  events_publisher_->publish(*parameter_event);
```

Comment:
It would be better if this kind of change were not needed.

Reply:
Do you mean that it would be better if the interface accept both pointer and reference

Reply:
Yeah, specifically it should continue to take a `shared_ptr`, either by reference or value, so that this code doesn't need to change.

Reply:
I think the API should be hard to misuse. Convenience to use is good to have but it shouldn't confuse the user.
- When you receive a `std::shared_ptr<>`, it mean that you will extend it lifetime (may be you will need to hold the reference of the object in another thread). Do you need it in this case? Remember `std::shared_ptr` is not free.
- `T const *` is good enough but `T const &` is better in this case, because the pointer will accept a `nullptr`, and in the function `publish` will throw an exception when the argument is `nullptr` in this case.

Reply:
What's the problem (i.e. how will it be misused?) with `const std::shared_ptr<T> &`? It neither implies extending the lifetime (though it could if needed) nor does it incur the cost of copying a shared pointer.

In the future we may want to make a copy of the shared pointer for use in the intra-process manager, so it's convenient to have in the future and right now, many people choose to create shared pointers to (using `std::make_shared`) messages for various reasons, e.g. like having a pool of reusable messages.

Reply:
`const std::shared_ptr<T> &` don't have lifetime problem, only `publish(std::shared_ptr<const MessageT> msg)` has problem. But if any time you have these pointer, you can make a call to `T const&`, so you will not need these interface. I think `*` is not a big typing problem.

and `const std::shared_ptr<T> &` can make user confuse that this function can modify their variable.

Reply:
What about `const std::shared_ptr<const T> & ptr`?

The goal here is to not be too disruptive to existing code and to allow people who are using shared pointers to messages to publish easily. I take the point that `*` is not hard to type, but it is a continuing cost, whereas us providing one more signature is also not a burden and a one time cost (for the most part).

So if we can come up with a signature that accurately reflects our intent and is not misleading then we should do that I think.